modules/extensions-support/src/main/java/org/mule/module/extension/internal/runtime/config/DynamicConfigurationInstanceProvider.java

```diff
@@ -0,0 +1,180 @@
+/*
+ * Copyright (c) MuleSoft, Inc.  All rights reserved.  http://www.mulesoft.com
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.module.extension.internal.runtime.config;
+
+import static org.mule.module.extension.internal.util.MuleExtensionUtils.asOperationContextAdapter;
+import org.mule.api.MuleEvent;
+import org.mule.api.MuleException;
+import org.mule.api.MuleRuntimeException;
+import org.mule.extension.introspection.Extension;
+import org.mule.extension.runtime.ConfigurationInstanceProvider;
+import org.mule.extension.runtime.ConfigurationInstanceRegistrationCallback;
+import org.mule.extension.runtime.ExpirationPolicy;
+import org.mule.extension.runtime.OperationContext;
+import org.mule.module.extension.internal.runtime.ExpirableContainer;
+import org.mule.module.extension.internal.runtime.resolver.ResolverSet;
+import org.mule.module.extension.internal.runtime.resolver.ResolverSetResult;
+
+import com.google.common.collect.ImmutableMap;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+/**
+ * A {@link ConfigurationInstanceProvider} which continuously evaluates the same
+ * {@link ResolverSet} and then uses the resulting {@link ResolverSetResult}
+ * to build an instance of a given type.
+ * <p/>
+ * Although each invocation to {@link #get(OperationContext)} is guaranteed to end up in an invocation
+ * to {@link #resolverSet#resolve(MuleEvent)}, the resulting {@link ResolverSetResult} might not end up
+ * generating a new instance. This is so because {@link ResolverSetResult} instances are put in a cache to
+ * guarantee that equivalent evaluations of the {@code resolverSet} return the same instance.
+ *
+ * @since 3.7.0
```

Comment:
Is correct to maintain 3.7.0 here when the class was move to a different package?